GPII-2296:  Initial commit of code, tests, and documentation for gpii-testem…

the-t-in-rtf opened the pull request:
See [GPII-2296](https://issues.gpii.net/browse/GPII-2296) for details.

the-t-in-rtf committed:
GPII-2296:  Initial commit of code, tests, and documentation for gpii-testem.

the-t-in-rtf committed:
GPII-2296:  Expanded usage instructions.

the-t-in-rtf committed:
GPII-2296: Tidied up the documentation based on the display on GitHub.

the-t-in-rtf committed:
GPII-2296: Fixed table formatting in lifecycle docs.

the-t-in-rtf committed:
GPII-2296: Minor tweaks to lifecycle docs.

the-t-in-rtf committed:
GPII-2296: Fixed lifecycle example to use `getTestemOptions()`.

the-t-in-rtf committed:
GPII-2296:  Updated lifecycle docs to remove inaccurate information about Fluid events and to provide clearer examples.

the-t-in-rtf committed:
GPII-2296: Fixed stray “unpinned” dependency version.

the-t-in-rtf committed:
NOJIRA: Added initial next-gen QI config file.

the-t-in-rtf committed:
GPII-2304:  Minor tweak to QI config to fix “incomplete” Firefox install.

the-t-in-rtf committed:
GPII-2296: Fixed coverage docs based on review on GitHub.

the-t-in-rtf committed:
GPII-2296: Improved cleanup on startup / shutdown.  Added tests of various failure modes.

the-t-in-rtf committed:
GPII-2296:  Updated docs to reflect the revised pathing and cleanup behavior.

the-t-in-rtf committed:
GPII-2296:  Fixed merge policy.

the-t-in-rtf committed:
FLUID-6140: Updated to hopefully pick up circular resolution fix.

the-t-in-rtf committed:
GPII-2296: Added non-source to instrumentation command so that config files, etc. are available during testing.

the-t-in-rtf committed:
GPII-2296:  Added special handling for Testem errors in “failure” tests.

amb26 commented:
Somehow the tests hang for me, after having shown the "Incomplete Coverage" Tests. The Chrome window comes up, and shows 2 passed tests out of 2, but does not dismiss.

amb26 commented on src/js/testem-component.js:
```diff
@@ -0,0 +1,531 @@
+        }
+    },
+    invokers: {
+        "handleTestemStart": {
```
I'd really like to see these embodied as namespaced listeners to an event rather than invokers. It seems quite likely that the "responsible" user might want to insert themselves into this workflow.
We can always embody the "failsafe" part of the workflow into a listener with priority "last" - this doesn't lead to any increased risk than the one you already have within gpii.testem.init - since if an exception is thrown during the processing of listeners to "onFixturesConstructed", the callback will be dangling in any case. And I think users of the design will appreciate there being Infusion events named by direct correspondence to the underlying testem ones.

the-t-in-rtf replied:
Just confirmed that the listeners are fired sequentially, so we would have a chain whose first activity is to save the callback to a member, and whose last is to call the callback.

the-t-in-rtf replied:
I have I hope addressed this comment in spirit.  The latest code uses `fluid.promise.makeTransformer` to fire two "pseudo events" when Testem starts up and shuts down.  Namespaced and prioritised listeners can register themselves to respond, and are executed in sequence.

Although I had originally considered storing the Testem callbacks and referring to those in IoC references, I ended up adding the Testem callback as both the `onReject` and `onResolve` handler for the sequence returned by `fluid.promise.makeTransformer`.  As failing to call the callbacks is a fundamental error that results in a hung process, I felt it was appropriate to hard-code this bit of wiring as an implicit and immovable last step in the chain.

amb26 replied:
I think that someone familiar with Infusion's idioms will still find the design of these functions puzzling and unidiomatic. Especially, "invoker that registers event listener" to me smells like a bit of an antipattern. The point of registering listeners declaratively is in order to make the design of an application clearer to authors - especially given that future tooling will expose this structure usefully.

The design of Infusion components actively promote ways in which misconfiguration may cause fundamental errors, at least in the opinion of their original author. I'd like to go one more round of promoting my original suggestion of implementing these functions as namespaced listeners plus a comment warning the user that overriding the one that triggers the callback without substituting an equivalent effect will cause a hung process. But this is something we can certainly discuss at a 1-to-1 next week - is your traditional Monday afternoon slot available? I can also do mornings now :)

amb26 commented on src/js/testem-component.js:
```diff
@@ -0,0 +1,531 @@
+    var testemRegexp = /testem-.+/;
+    var togo = fluid.promise();
+
+    try {
```
The bodies of functions in the file seem to be written to a kind of "recurring pattern" in that they do something fallible, catch an exception and then convert it to a promise rejection. I am speculating on what it might mean for this kind of workflow to be factored out into a common utility. I can't see any clear way of doing this that offers any particular benefits, but thought that I would just drop this thought in here for us to ponder about it.

the-t-in-rtf replied:
We pondered for a bit, and agreed that we would focus on converting the existing hodgepodge of events, callbacks, and promises into pseudo-event chains for this pass.

the-t-in-rtf committed:
GPII-2311: Use the new body parser options to raise the limit of data that can be posted to the coverage receiver.

the-t-in-rtf committed:
GPII-2311: Fixed typo in body parser options.

the-t-in-rtf committed:
GPII-2296:  Added support for bodyParserOptions to the url encoded middleware wrapper as well.

the-t-in-rtf committed:
GPII-2296: Updated to pick up the new version of express that allows configuring the url encoding middleware.

the-t-in-rtf committed:
GPII-2296: Refined coverage harness to avoid using Testem’s proxy.

the-t-in-rtf committed:
NOJIRA: Updated to latest QI configuration file format.

the-t-in-rtf committed:
Add browser, test fixture details to coverage filenames.

the-t-in-rtf committed:
NOJIRA: Sketched out “git diff + git clone” workaround for Vagrant tests.

the-t-in-rtf committed:
GPII-2296: Updated dependencies, including released version of gpii-express.

the-t-in-rtf committed:
NOJIRA:  Continued testing workarounds for QI Windows VMs and tests.

the-t-in-rtf committed:
GPII-2296:  Moved `mkdirp` to the main dependencies to avoid problems running tests on machines without that installed globally.

the-t-in-rtf commented:
In recent work on the gpii-pouchdb pull, we've encountered hangs in both ci and manual Testem runs.  In looking into this, I noticed that if I close the browser, the tests continue, but fail because there is no coverage data.  I noticed that the tests with no coverage run in all browsers with no problem.  This to me suggests that the problem is that the browser is not calling the Testem.afterTests callback once the coverage is reported.  I will have to do some testing to see if this is a firewall issue, et cetera.

the-t-in-rtf commented:
In the final coverage report from the aborted run, I also notice 0% branch coverage for the "receiver" middleware.  This to me suggests that the receive is never getting any requests.  I will confirm shortly whether it starts up, is reachable, etc.

the-t-in-rtf committed:
GPII-2296:  Ensure Testem callback will also fire when no coverage data is available.

the-t-in-rtf commented:
I have confirmed that no coverage data is being sent because the source code is not instrumented on Windows.  I'm troubleshooting this now.

the-t-in-rtf committed:
GPII-2296: Added cwd to process exec options to hopefully address windows instrumentation issues.

the-t-in-rtf commented:
Late yesterday I confirmed that the source code was in fact instrumented, this appears to be a problem with the [routes](https://github.com/testem/testem/blob/master/docs/config_file.md#config-level-options) used to replace the source path with the instrumented code.  I am investigating further now.

the-t-in-rtf committed:
GPII-2296:  Updated routing logic to avoid problems in Windows.  Also improved logging for istanbul commands.

the-t-in-rtf commented:
I have fixed the windows routing errors.  In my testing I noticed shutdown hangs when running Testem in regular (non-CI) mode.  I am investigating that and will confirm once this is safe to review.

the-t-in-rtf committed:
GPII-2296:  Downgrade Testem version to avoid hangs when running in non-ci mode.

the-t-in-rtf committed:
GPII-2296:  Updated tests to use “our” copy of Testem rather than a global one, so that we can ensure we use a working version (1.13.0).

the-t-in-rtf committed:
GPII-2296: Updated README to reflect the fact that tests now run with a “local” copy of Testem rather than a global one.

the-t-in-rtf committed:
GPII-2296: Updated docs to prefer “local” copies of Testem and to point out the bug with versions of Testem higher than 1.13.0

the-t-in-rtf commented:
There appears to be [a bug in version 1.14.0 and higher of Testem](https://github.com/testem/testem/issues/1075) that causes hangs when exiting in non-ci mode.  I have downgrade the package version of testem and ensured that the tests use a "local" copy of Testem.  I also updated the docs to:

1. mention the bug
2. use "local" copies of Testem in the example commands
3. Indicate that Testem does not need to be installed globally to run the tests in the package itself.

@amb26, this should now be ready for review again.

the-t-in-rtf committed:
GPII-2296: Updated to use gpii-express 1.0.5 `middlewareOptions` convention instead of previous `bodyParserOptions`.

the-t-in-rtf committed:
GPII-2296: Split out inline handlebars “invoker” template into a separate file to aid in debugging problems.

the-t-in-rtf committed:
Trivial marketing change to allow me to cut a new dev release.

the-t-in-rtf committed:
NOJIRA:  Added hooks to force QUnit to start only after the coverage client has hooked into Testem’s `afterTests` event.

the-t-in-rtf committed:
GPII-2296: Updated documentation to reflect coverage sender dependencies, and the fact that our focus is QUnit.

the-t-in-rtf commented:
In further testing with Infusion, I added code to ensure that tests are not run before the coverage sender is set up.  I also updated the examples to reflect the full dependencies of the coverage sender.

the-t-in-rtf committed:
FLUID-6140:  Updated to use a version of gpii-express that matches “ours”, to resolve “dedupe” errors when running in Windows VMs.

the-t-in-rtf committed:
GPII-2303: Refined QI config file based on recent investigations. Tests now pass.

the-t-in-rtf committed:
GPII-2303: Workaround for fluid.require errors.

the-t-in-rtf commented:
Just to summarise, now that the tests run in a Windows VM, the primary work remaining is to refactor to use namespaced event listeners, as discussed [here](https://github.com/GPII/gpii-testem/pull/1#discussion_r103993403).

the-t-in-rtf committed:
NOJIRA:  Fixed QI config problem (v: not visible from do.ps1).

the-t-in-rtf committed:
FLUID-6149: Updated to dev release of infusion and gpii-express to pick up fluid.require fix.

the-t-in-rtf committed:
GPII-2296:  Updated Testem component to use transforming promise chains for Testem’s startup and shutdown.

the-t-in-rtf commented:
There are still some issues with running the tests using the QI plugin, this is otherwise ready for review.

the-t-in-rtf committed:
GPII-2296: Initial working QI config.  Also improved calls to istanbul to remove the need to have it installed globally.

the-t-in-rtf committed:
GPII-2296:  Cleaned up documentation following recent changes.  Renamed `testemRoot` to reduce confusion.

the-t-in-rtf commented:
@amb26, the QI config now works as well.   I have updated the docs to reflect recent changes as well.  We need to review the use of fluid.promise.makeTransform later today.

the-t-in-rtf committed:
GPII-2296: Refactored event and timeout code into well-tested static functions.

the-t-in-rtf commented:
@amb26, I have completed the refactor we discussed, the event and timeout handling are now static functions with their own tests.  This should be ready for further review.

the-t-in-rtf committed:
NOJIRA:  Disabled QUnit autostart mechanism based on errors in other packages.

the-t-in-rtf committed:
GPII-2296: Added support for common "testDef" format and tests for same.

the-t-in-rtf commented:
In a side discussion on chat, @amb26 and I agreed that I would create two "test loaders" to help ensure that the same tests are run both in qunit-composite and testem.  This PR now reflects the gpii-testem part of that work, and is ready for further review.

the-t-in-rtf commented:
In using this package with universal and infusion, I started encountering more and more "browser disconnect" errors.  I tried a lot of things before I was able to get both to run consistently in Vagrant, including:

1. Increasing the amount of RAM available to the VM.
2. Increasing the browser timeouts beyond Testem's defaults.
3. Adding memory tuning options for Chrome.
4. Bypassing qunit-chrome.

After painstaking troubleshooting (emphasis on "pain"), I have determined that only the second and third step are truly necessary.  I can make these the defaults for the base grade, and remove the "testDefs" support here, in the infusion PR, and in the universal PR.

the-t-in-rtf committed:
Revert "GPII-2296: Added support for common "testDef" format and tests for same."

This reverts commit 70e12cf5032856a37f47462ab5d8bbb3ab178469.

the-t-in-rtf committed:
GPII-2296:  Added options to minimize "browser disconnect" errors by default.

the-t-in-rtf commented:
The latest commits added the minimum required fixes to avoid "browser disconnect" errors to the defaults for the `gpii.testem` grade.   This is confirmed working in both the universal and infusion PRs also in progress.  @amb26, this is ready for review.

the-t-in-rtf committed:
GPII-2296: Updated to use newest Windows 10 VM for QI tests.

the-t-in-rtf commented:
With the latest Windows 10 VM, the tests now run using the QI as well, i.e. `vagrant up` and `vagrant ci test` are enough to run the tests in Windows, including generating a coverage report.

the-t-in-rtf committed:
GPII-2380: Updated to "root safe dedupe" version of kettle and infusion, including in dependencies..

the-t-in-rtf committed:
NOJIRA: Removed unneeded `gpii-handlebars` dependency.

the-t-in-rtf committed:
GPII-2380: Updated to "root safe dedupe" version of `gpii-express` that includes safe version of `kettle`.

the-t-in-rtf committed:
Added initial GitLab CI configuration

the-t-in-rtf committed:
GPII-2380: Updated to "root safe dedupe" release version of gpii-express.

the-t-in-rtf committed:
GPII-2380: Updated to "root safe dedupe" release version of gpii-express.

the-t-in-rtf committed:
GPII-2396: Added cleanup of (empty) user data directory...

the-t-in-rtf committed:
GPII-2396: Greatly improved cleanup of temporary content following test runs.

the-t-in-rtf commented:
@amb26, although this is now in a pretty good state, I'd like to wait to merge and release until we hear back regarding [the testem "browser_args" pull](https://github.com/testem/testem/pull/1120).

the-t-in-rtf commented:
The testem pull has been merged, just waiting for a release.

the-t-in-rtf committed:
GPII-2400: Updated to released version of testem with better support for arrays of browser arguments.

the-t-in-rtf committed:
NOJIRA: Removed (now) unnecessary proactive upgrade of firefox, which was previously required to ensure a working firefox in the Vagrant VM.

the-t-in-rtf commented:
@amb26, I have updated the branch to use the released version of testem that includes the fix for GPII-2400, and confirmed that all tests pass in the Vagrant VM.  This is now ready for further review.

the-t-in-rtf committed:
NOJIRA: Updated tests to run on IE.

the-t-in-rtf committed:
NOJIRA: Updated Gitlab config per feedback from Avtar...

amb26 commented on docs/testem-component.md:
```diff
@@ -0,0 +1,87 @@
+
+## `{gpii.testem}.getTestemOptions()`
+
+An invoker which exposes the final Testem options in a way that Testem can read from a configuration file.  See
```
The language "in a way that Testem can read" doesn't quite make sense to me. Is it meant to express that the options take the same form that Testem might read from a configuration file, or that it will do so?

the-t-in-rtf replied:
I will update this to clarify, basically when using the `testem.js` variant, the options have to be exported via `module.exports`.

the-t-in-rtf replied:
I have submitted alternate wording with links to the relevant Testem docs.  Please review.

the-t-in-rtf committed:
GPII-2296: Clarified the purpose of the `getTestemOptions` invoker.

the-t-in-rtf committed:
NOJIRA:  Updated dependency versions per output of `yarn outdated`.

the-t-in-rtf committed:
NOJIRA:  Updated `nyc` dependency.

the-t-in-rtf committed:
NOJIRA: Updated docs to reflect the actual dependencies of the coverage client itself, based on feedback in other reviews.

the-t-in-rtf commented:
In working on the ["code coverage" universal PR](https://github.com/GPII/universal/pull/508#issuecomment-307817581), @amb26 pointed out one unneeded dependency.  I discovered two more while working to address his feedback.  I have updated the docs to reflect the minimum required dependencies for the code coverage client itself.

amb26 commented on tests/unit-tests.js:
```diff
@@ -0,0 +1,158 @@
+    timeoutPromise.then(
+        function () {
+            jqUnit.start();
+            fluid.fail("The wrapped promise should not have been resolved.")
```
I am getting widespread linting errors on this pull when running "grunt lint". Some of them seem to reflect genuine errors, such as a missing semicolon on this line - others seem to reflect some kind of linting misconfiguration, e.g. "use the function form of use strict" which shouldn't be an error we receive by our default rules.

amb26 replied:
When the linting is fixed, I imagine we should also exclude gpii-testem\tests\testem-fixtures\instrumented from linting

the-t-in-rtf replied:
All linting errors fixed, and the instrumented directory has been excluded.  I also fixed my precommit hook, which should have caught this.

the-t-in-rtf committed:
GPII-2296: Cleaned up various previously cloaked lint errors.

the-t-in-rtf committed:
GPII-2296: Cleaned up various previously cloaked lint errors.

amb26 commented:
Dear @the-t-in-rtf - thanks for your skilled and dedicated long-term work on this massive pull request, implementing functionality which will be crucial across all of our GPII and Fluid projects. I've taken the liberty of rolling and tagging the release - many congratulations, you are a star!

the-t-in-rtf commented:
Thanks for all the help getting this one ready @amb26, I'm glad to get it out there.  I updated the version in master to be forward-facing, and have updated dependent PRs to use the released version.

the-t-in-rtf commented:
I also added release notes to the version on GitHub.
